pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,212 @@
+    }
+    _semaphore = new AdjustableSemaphore(preprocessConcurrency, true);
+    LOGGER.info("Created semaphore with total permits: {}, available permits: {}", totalPermits(),
+        availablePermits());
+  }
+
+  public synchronized void resetThrottling() {
+    LOGGER.info("Reset throttling threshold for segment preprocess concurrency, total permits: {}, available "
+            + "permits: {}", totalPermits(), availablePermits());
+    _relaxThrottling = false;
+    _semaphore.setPermits(_maxPreprocessConcurrency);
+    LOGGER.info("Reset throttling completed, new concurrency: {}, total permits: {}, available permits: {}",
+        _maxPreprocessConcurrency, totalPermits(), availablePermits());
+  }
+
+  @Override
+  public synchronized void onChange(Set<String> changedConfigs, Map<String, String> clusterConfigs) {
+    if (clusterConfigs == null || clusterConfigs.isEmpty()) {
+      LOGGER.info("Skip updating SegmentPreprocessThrottler configs with empty clusterConfigs");
+      return;
+    }
+
+    if (changedConfigs == null || changedConfigs.isEmpty()) {
+      LOGGER.info("Skip updating SegmentPreprocessThrottler configs with unchanged clusterConfigs");
+      return;
+    }
+
+    LOGGER.info("Updating SegmentPreprocessThrottler configs with latest clusterConfigs");
+    handleMaxPreprocessConcurrencyChange(changedConfigs, clusterConfigs);
+    handleMaxPreprocessConcurrencyBeforeServingQueriesChange(changedConfigs, clusterConfigs);
+    LOGGER.info("Updated SegmentPreprocessThrottler configs with latest clusterConfigs, total permits: {}",
+        totalPermits());
+  }
+
+  private void handleMaxPreprocessConcurrencyChange(Set<String> changedConfigs, Map<String, String> clusterConfigs) {
+    if (!changedConfigs.contains(CommonConstants.Helix.CONFIG_OF_MAX_SEGMENT_PREPROCESS_PARALLELISM)) {
+      LOGGER.info("changedConfigs list indicates maxPreprocessConcurrency was not updated, skipping updates");
+      return;
+    }
+
+    String configName = CommonConstants.Helix.CONFIG_OF_MAX_SEGMENT_PREPROCESS_PARALLELISM;
+    String defaultConfigValue = CommonConstants.Helix.DEFAULT_MAX_SEGMENT_PREPROCESS_PARALLELISM;
+    String maxParallelSegmentPreprocessesStr = clusterConfigs.getOrDefault(configName, defaultConfigValue);
+    int maxPreprocessConcurrency = Integer.parseInt(maxParallelSegmentPreprocessesStr);
+
+    if (maxPreprocessConcurrency == _maxPreprocessConcurrency) {
+      LOGGER.info("No ZK update for maxPreprocessConcurrency {}, total permits: {}", _maxPreprocessConcurrency,
+          totalPermits());
+      return;
+    }
+
+    if (maxPreprocessConcurrency <= 0) {
+      LOGGER.warn("Invalid maxPreprocessConcurrency set: {}, not making change, fix config and try again",
+          maxPreprocessConcurrency);
+      return;
+    }
+
+    LOGGER.info("Updated maxPreprocessConcurrency from: {} to: {}", _maxPreprocessConcurrency,
```

Comment:
Move this log after actually set the permits, in case the `setPermits` failed

Reply:
done